paddle/fluid/operators/py_func_op.cc

```diff
@@ -0,0 +1,285 @@
+                   "Input(X) or Output(Out) must exist");
+    PADDLE_ENFORCE_GE(ctx->Attrs().Get<int>(kForwardPythonCallableId), 0,
+                      "Function id cannot be less than 0");
+
+    // Transverse all outputs
+    // If name of any output ends with @GRAD,
+    // set its shape, dtype, lod_level, type to be the same as
+    // the correponding forward variable
+    auto *op = boost::get<const framework::OpDesc *>(ctx->GetOp());
+    auto *block = op->Block();
+    const std::string kGradVarSuffix = framework::kGradVarSuffix;
+    auto out_vars = ctx->GetOutputVarPtrs("Out");
+    for (auto &out_var : out_vars) {
+      auto *out_var_desc = boost::get<framework::VarDesc *>(out_var);
+      if (out_var_desc == nullptr) {
+        continue;
+      }
+      auto out_name = out_var_desc->Name();
+      if (out_name == framework::kEmptyVarName ||
+          out_name.size() <= kGradVarSuffix.size()) {
+        continue;
+      }
+
+      size_t len = out_name.size() - kGradVarSuffix.size();
+      if (out_name.substr(len) == kGradVarSuffix) {
+        auto fwd_var_name = out_name.substr(0, len);
+        auto *in_var_desc = block->FindVarRecursive(fwd_var_name);
+        PADDLE_ENFORCE_NOT_NULL(in_var_desc, "Forward variable %s not found",
+                                fwd_var_name);
+        VLOG(10) << "Infer shape of Output(" << out_name << ") as Input("
+                 << in_var_desc->Name() << ")";
+        out_var_desc->SetShape(in_var_desc->GetShape());
+        out_var_desc->SetDataType(in_var_desc->GetDataType());
+        out_var_desc->SetLoDLevel(in_var_desc->GetLoDLevel());
+        out_var_desc->SetType(in_var_desc->GetType());
+      }
+    }
+  }
+};
+
+class PyFuncOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "Inputs of py_func op.").AsDuplicable();
+    AddOutput("Out", "Outputs of py_func op").AsDuplicable();
+    AddAttr<int>(kForwardPythonCallableId,
+                 "Index of registered forward Python function.")
+        .SetDefault(0);
+    AddAttr<int>(kBackwardPythonCallableId,
+                 "Index of registered backward Python function.")
+        .SetDefault(-1);
+    AddAttr<std::vector<std::string>>(kPyFuncBackwardSkipVars,
+                                      "Unused forward in/out in backward op")
+        .SetDefault(std::vector<std::string>());
+    AddComment(R"DOC("PyFunc Op")DOC");
+  }
+};
+
+class PyFuncOpGradDescMaker : public framework::GradOpDescMakerBase {
+ public:
+  using framework::GradOpDescMakerBase::GradOpDescMakerBase;
+
+  std::vector<std::unique_ptr<framework::OpDesc>> operator()() const override {
+    auto &fwd_attrs = Attrs();
+    // no backward op when backward_id is less than 0
+    if (boost::get<int>(fwd_attrs.at(kBackwardPythonCallableId)) < 0) {
+      return {};
+    }
+
+    std::unique_ptr<framework::OpDesc> grad_op(new framework::OpDesc());
+    grad_op->SetType("py_func");
```

Comment:
why not py_func_grad?

Reply:
As comment in https://github.com/PaddlePaddle/Paddle/pull/14836#discussion_r242010867, support high order derivatives.